docs/blocks-storage/migrate-from-chunks-to-blocks.md

```diff
@@ -0,0 +1,275 @@
+
+## Step 1: Preparation
+
+Before starting the migration of ingesters, we need to prepare other services.
+
+### Querier and Ruler
+
+_Everything discussed for querier applies to ruler as well, since it shares querier configuration – options prefix is `querier` even when used by ruler._
+
+Querier and ruler need to be reconfigured as follow:
+
+- `-querier.second-store-engine=blocks`
+- `-querier.query-store-after=0`
+- `-querier.ingester-streaming=false`
+
+#### `-querier.second-store-engine=blocks`
+
+Querier (and ruler) needs to be reconfigured to query both chunks storage and blocks storage at the same time.
+This is achieved by using `-querier.second-store-engine=blocks` option, and providing querier with full blocks configuration, but keeping "primary" store set to `-store.engine=chunks`.
+
+#### `-querier.query-store-after=0`
+
+Querier (and ruler) has an option `-querier.query-store-after` to query store only if query hits data older than some period of time.
+For example, if ingesters keep 12h of data in memory, there is no need to hit the store for queries that only need last 1h of data.
+During the migration, this flag needs to be set to 0, to make queriers always consult the store when handling queries.
+The reason is that after chunks ingesters shutdown, they can no longer return chunks from memory.
+
+#### `-querier.ingester-streaming=false`
+
+Querier (and ruler) has a [bug](https://github.com/cortexproject/cortex/issues/2935) and doesn't properly
+merge streamed results from chunks and blocks-based ingesters. Instead it only returns data from blocks- instesters.
+To avoid this problem, we need to temporarily disable this feature by setting `-querier.ingester-streaming=false`.
+After migration is complete (i.e. all ingesters are running blocks only), this can be turned back to true, which is the default value.
+
+### Query-frontend
+
+Query-frontend needs to be reconfigured as follow:
+
+- `-querier.parallelise-shardable-queries=false`
+
+#### `-querier.parallelise-shardable-queries=false`
+
+Query frontend has an option `-querier.parallelise-shardable-queries` to split some incoming queries into multiple queries based on sharding factor used in v11 schema of chunk storage.
+As the description implies, it only works when using chunks storage.
+During and after the migration to blocks (and also after possible rollback), this option needs to be disabled otherwise query-frontend will generate queries that cannot be satisfied by blocks.
+
+### Compactor and Store-gateway
+
+[Compactor](./compactor.md) and [store-gateway](./store-gateway.md) services should be deployed and successfully up and running before migrating ingesters.
+
+## Step 2: Ingesters migration
+
+We have developed a script available in Cortex [`tools/migrate-ingester-statefulsets.sh`](https://github.com/cortexproject/cortex/blob/master/tools/migrate-ingester-statefulsets.sh) to migrate ingesters between two StatefulSets, shutting down ingesters one by one.
```

Comment:
The missing part here is that it assumes that the new StatefulSet has been created with the right configuration and the number of replicas set to 0. I believe this should be covered adding a section to the preparation phase (above).